File: src/workflows/inputs.ts
import { shellArgv, spawnCapture } from "../runner/shell";
import { bail, type FlatStep, type InputSpec } from "./types";
import type { RawWorkflow } from "./parse";
import { isBuiltin } from "./substitute";
import { AGENT_NAME_RE, stepOutNames, stepReferencedNames } from "./steps";
import { AGENTS_BUILTIN, normalizeInput } from "./input-normalize";

export { AGENT_NAME_RE };

const OPTIONS_CMD_TIMEOUT_MS = 5_000;

/** True when `command` references `$HWF_<name>` or `$HWF_INPUT_<name>` exactly. */
function commandUsesEnv(command: string, name: string): boolean {
  for (const prefix of [`HWF_${name}`, `HWF_INPUT_${name}`]) {
    let from = 0;
    while (from <= command.length) {
      const i = command.indexOf(prefix, from);
      if (i === -1) break;
      const after = command[i + prefix.length];
      if (after === undefined || !/[A-Za-z0-9_]/.test(after)) return true;
      from = i + prefix.length;
    }
  }
  return false;
}

async function resolveOptionLines(
  file: string,
  inputName: string,
  command: string,
  repoRoot: string,
): Promise<string[]> {
  const result = await spawnCapture(shellArgv(command), {
    cwd: repoRoot,
    timeoutMs: OPTIONS_CMD_TIMEOUT_MS,
  });
  if (result.timedOut) {
    bail(
      file,
      undefined,
      `inputs.${inputName}`,
      `options command timed out after ${OPTIONS_CMD_TIMEOUT_MS / 1000}s`,
    );
  }
  if (result.exitCode !== 0) {
    const detail = result.stderr.trim() || `exit ${result.exitCode}`;
    bail(file, undefined, `inputs.${inputName}`, `options command failed: ${detail}`);
  }
  const lines = result.stdout
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0);
  if (lines.length === 0) {
    bail(file, undefined, `inputs.${inputName}`, "options command produced no choices");
  }
  return lines;
}

export async function resolveInputs(
  file: string,
  raw: RawWorkflow,
  agents: Set<string>,
  repoRoot: string,
  resolveDynamic = true,
): Promise<InputSpec[]> {
  const specs: InputSpec[] = [];
  for (const [name, value] of Object.entries(raw.inputs ?? {})) {
    if (isBuiltin(name)) {
      bail(file, undefined, `inputs.${name}`, `name shadows builtin '{${name}}'`);
    }
    const input = normalizeInput(file, name, value);
    let options: string[] | undefined;
    let dynamicOptions = false;
    if (input.options !== undefined) {
      if (Array.isArray(input.options)) {
        options = input.options;
      } else if (input.options === AGENTS_BUILTIN) {
        if (agents.size === 0) {
          bail(file, undefined, `inputs.${name}`, "options: agents but no agents configured");
        }
        options = [...agents];
      } else if (resolveDynamic) {
        options = await resolveOptionLines(file, name, input.options, repoRoot);
      } else {
        dynamicOptions = true;
      }
    }
    if (options && input.default !== undefined && !options.includes(input.default)) {
      bail(file, undefined, `inputs.${name}`, `default '${input.default}' not in options`);
    }
    specs.push({
      name,
      label: input.label ?? name,
      desc: input.desc,
      options,
      ...(dynamicOptions ? { dynamicOptions: true } : {}),
      default: input.default,
    });
  }
  return specs;
}

function checkAgentInput(file: string, idx: number, spec: InputSpec, agents: Set<string>): void {
  if (spec.dynamicOptions) return;
  if (!spec.options) {
    bail(file, idx + 1, "agent", `input '${spec.name}' needs options: to be used as agent`);
  }
  for (const option of spec.options) {
    if (!agents.has(option)) {
      bail(file, idx + 1, "agent", `input '${spec.name}' option '${option}' is not a config agent`);
    }
  }
}

function checkIncludeNames(
  file: string,
  step: FlatStep,
  stepIndex: number,
  inputs: Map<string, InputSpec>,
  agents: Set<string>,
  knownNames: Set<string>,
  used: Set<string>,
  markInputUse: boolean,
): void {
  if (step.action.kind !== "include") return;
  const action = step.action;
  for (const name of stepReferencedNames(step)) {
    if (isBuiltin(name)) continue;
    if (!knownNames.has(name) && !inputs.has(name)) {
      bail(file, stepIndex, "with", `unknown name '{${name}}'`);
    }
    if (inputs.has(name) && markInputUse) used.add(name);
  }
  const childKnown = new Set<string>([
    ...Object.keys(action.with),
    ...Object.keys(action.defaults),
  ]);
  checkStepNames(file, new Map(), action.steps, agents, childKnown, used, false);
  for (const n of action.exportedOuts) knownNames.add(n);
}

function checkOneReferencedName(
  file: string,
  step: FlatStep,
  stepIndex: number,
  name: string,
  inputs: Map<string, InputSpec>,
  knownNames: Set<string>,
  used: Set<string>,
  markInputUse: boolean,
): void {
  if (name === "item" || name === "index") {
    if (!step.for) {
      bail(file, stepIndex, undefined, `{${name}} requires for:`);
    }
    return;
  }
  if (name === "attempt") {
    if (!step.retry) {
      bail(file, stepIndex, undefined, `{attempt} requires retry:`);
    }
    return;
  }
  if (isBuiltin(name)) return;
  if (knownNames.has(name)) {
    if (inputs.has(name) && markInputUse) used.add(name);
    return;
  }
  if (inputs.has(name)) {
    if (markInputUse) used.add(name);
    return;
  }
  bail(file, stepIndex, undefined, `unknown name '{${name}}'`);
}

function markRunEnvInputUses(
  step: FlatStep,
  inputs: Map<string, InputSpec>,
  used: Set<string>,
  markInputUse: boolean,
): void {
  if (step.action.kind !== "run" || step.action.payload.form === "argv") return;
  const command = step.action.payload.command;
  for (const name of inputs.keys()) {
    if (commandUsesEnv(command, name) && markInputUse) used.add(name);
  }
}

function checkAgentStepName(
  file: string,
  step: FlatStep,
  idx: number,
  stepIndex: number,
  inputs: Map<string, InputSpec>,
  agents: Set<string>,
  used: Set<string>,
  markInputUse: boolean,
): void {
  if (step.action.kind !== "agent") return;
  const m = AGENT_NAME_RE.exec(step.action.agent);
  if (!m || isBuiltin(m[1]!)) return;
  const spec = inputs.get(m[1]!);
  if (!spec) {
    bail(file, stepIndex, "agent", `unknown name '{${m[1]}}'`);
  }
  if (markInputUse) used.add(m[1]!);
  checkAgentInput(file, idx, spec, agents);
}

function checkAsAndOutNames(
  file: string,
  step: FlatStep,
  stepIndex: number,
  inputs: Map<string, InputSpec>,
  knownNames: Set<string>,
): void {
  if (step.as) {
    if (isBuiltin(step.as) || inputs.has(step.as) || knownNames.has(step.as)) {
      bail(file, stepIndex, "as", `name '${step.as}' collides with an existing name`);
    }
  }
  for (const name of stepOutNames(step)) {
    if (isBuiltin(name)) {
      bail(file, stepIndex, "out", `name shadows a builtin '{${name}}'`);
    }
    if (inputs.has(name) || knownNames.has(name)) {
      bail(file, stepIndex, "out", `name '${name}' collides with an existing name`);
    }
    knownNames.add(name);
  }
}

function checkStepNames(
  file: string,
  inputs: Map<string, InputSpec>,
  steps: FlatStep[],
  agents: Set<string>,
  knownNames: Set<string>,
  used: Set<string>,
  markInputUse: boolean,
): void {
  let idx = 0;
  for (const step of steps) {
    const stepIndex = idx + 1;
    if (step.action.kind === "include") {
      checkIncludeNames(file, step, stepIndex, inputs, agents, knownNames, used, markInputUse);
      idx++;
      continue;
    }

    for (const name of stepReferencedNames(step)) {
      checkOneReferencedName(file, step, stepIndex, name, inputs, knownNames, used, markInputUse);
    }

    markRunEnvInputUses(step, inputs, used, markInputUse);
    checkAgentStepName(file, step, idx, stepIndex, inputs, agents, used, markInputUse);
    checkAsAndOutNames(file, step, stepIndex, inputs, knownNames);
    idx++;
  }
}

export function checkInputRefs(
  file: string,
  inputs: Map<string, InputSpec>,
  steps: FlatStep[],
  agents: Set<string>,
  knownNames: Set<string>,
): Set<string> {
  const used = new Set<string>();
  checkStepNames(file, inputs, steps, agents, knownNames, used, true);
  return used;
}

File: rust/src/workflow/inputs.rs
//! Input resolution + `checkInputRefs`. Port of `src/workflows/inputs.ts`.

use std::collections::HashSet;
use std::path::Path;
use std::time::Duration;

use super::errors::{InputSpec, WorkflowLoadError, positioned};
use super::placeholder::{agent_input_ref, params_input_refs, text_input_refs};
use super::steps::FlatStep;
use super::types::{RawOptions, RawWorkflow};
use crate::runner::spawn::{self, SpawnOpts};

const OPTIONS_CMD_TIMEOUT: Duration = Duration::from_millis(5_000);
const AGENTS_BUILTIN: &str = "agents";

/// Options-command capture via the shared helper in `crate::runner::spawn`
/// (`sh -c`, no stdin, inherited env).
fn spawn_capture(command: &str, cwd: &Path, timeout: Duration) -> std::io::Result<spawn::Capture> {
    let argv = vec!["sh".to_string(), "-c".to_string(), command.to_string()];
    spawn::spawn_capture(
        &argv,
        &SpawnOpts {
            cwd,
            stdin: None,
            env: &[],
            timeout,
        },
    )
}

/// `resolveOptionLines`: run the options command, map timeout/nonzero/empty
/// to the pinned load errors.
fn resolve_option_lines(
    file: &str,
    input_name: &str,
    command: &str,
    repo_root: &Path,
) -> Result<Vec<String>, WorkflowLoadError> {
    let key = format!("inputs.{input_name}");
    let result = spawn_capture(command, repo_root, OPTIONS_CMD_TIMEOUT).map_err(|e| {
        WorkflowLoadError(positioned(
            file,
            None,
            Some(&key),
            &format!("options command failed: {e}"),
        ))
    })?;
    if result.timed_out {
        return Err(WorkflowLoadError(positioned(
            file,
            None,
            Some(&key),
            &format!(
                "options command timed out after {}s",
                OPTIONS_CMD_TIMEOUT.as_secs()
            ),
        )));
    }
    if result.exit_code != 0 {
        let detail = result.stderr.trim();
        let detail = if detail.is_empty() {
            format!("exit {}", result.exit_code)
        } else {
            detail.to_string()
        };
        return Err(WorkflowLoadError(positioned(
            file,
            None,
            Some(&key),
            &format!("options command failed: {detail}"),
        )));
    }
    let lines: Vec<String> = result
        .stdout
        .split('\n')
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .map(str::to_string)
        .collect();
    if lines.is_empty() {
        return Err(WorkflowLoadError(positioned(
            file,
            None,
            Some(&key),
            "options command produced no choices",
        )));
    }
    Ok(lines)
}

/// `resolveInputs`: declaration order, `agents` builtin, dynamic option
/// commands (executed only when `resolve_dynamic`), default membership.
///
/// The `agents` builtin expands sorted: TS preserves config insertion order,
/// but the Rust config port (`BTreeMap`) is sorted, so sorted matches what
/// callers here actually hold.
///
/// # Errors
/// `WorkflowLoadError` for the agents-sentinel-without-agents, options-command
/// failures, or a default outside the resolved options.
pub fn resolve_inputs(
    file: &str,
    raw: &RawWorkflow,
    agents: &HashSet<String>,
    repo_root: &Path,
    resolve_dynamic: bool,
) -> Result<Vec<InputSpec>, WorkflowLoadError> {
    let mut specs = Vec::with_capacity(raw.input_order.len());
    for name in &raw.input_order {
        let input = raw
            .inputs
            .as_ref()
            .and_then(|inputs| inputs.get(name))
            .expect("input_order entries exist in inputs");
        let key = format!("inputs.{name}");
        let mut options: Option<Vec<String>> = None;
        match &input.options {
            None => {}
            Some(RawOptions::Choices(choices)) => options = Some(choices.clone()),
            Some(RawOptions::Command(command)) if command == AGENTS_BUILTIN => {
                if agents.is_empty() {
                    return Err(WorkflowLoadError(positioned(
                        file,
                        None,
                        Some(&key),
                        "options: agents but no agents configured",
                    )));
                }
                let mut expanded: Vec<String> = agents.iter().cloned().collect();
                expanded.sort();
                options = Some(expanded);
            }
            Some(RawOptions::Command(command)) => {
                if resolve_dynamic {
                    options = Some(resolve_option_lines(file, name, command, repo_root)?);
                }
            }
        }
        if let (Some(options), Some(default)) = (&options, &input.default) {
            if !options.contains(default) {
                return Err(WorkflowLoadError(positioned(
                    file,
                    None,
                    Some(&key),
                    &format!("default '{default}' not in options"),
                )));
            }
        }
        let dynamic_options = matches!(&input.options, Some(RawOptions::Command(command)) if command != AGENTS_BUILTIN)
            && !resolve_dynamic;
        specs.push(InputSpec {
            name: name.clone(),
            label: input.label.clone().unwrap_or_else(|| name.clone()),
            options,
            dynamic_options,
            default: input.default.clone(),
        });
    }
    Ok(specs)
}

/// `checkAgentInput`: an input driving `agent:` must resolve to config agents.
fn check_agent_input(
    file: &str,
    idx: usize,
    spec: &InputSpec,
    agents: &HashSet<String>,
) -> Result<(), WorkflowLoadError> {
    if spec.dynamic_options {
        return Ok(());
    }
    let Some(options) = &spec.options else {
        return Err(WorkflowLoadError(positioned(
            file,
            Some(idx + 1),
            Some("agent"),
            &format!("input '{}' needs options: to be used as agent", spec.name),
        )));
    };
    for option in options {
        if !agents.contains(option) {
            return Err(WorkflowLoadError(positioned(
                file,
                Some(idx + 1),
                Some("agent"),
                &format!(
                    "input '{}' option '{}' is not a config agent",
                    spec.name, option
                ),
            )));
        }
    }
    Ok(())
}

/// `require` in TS: look up a declared input, mark it used.
fn require<'a>(
    file: &str,
    inputs: &'a [InputSpec],
    name: &str,
    idx: usize,
    key: Option<&str>,
    used: &mut HashSet<String>,
) -> Result<&'a InputSpec, WorkflowLoadError> {
    let Some(spec) = inputs.iter().find(|spec| spec.name == name) else {
        return Err(WorkflowLoadError(positioned(
            file,
            Some(idx + 1),
            key,
            &format!("undeclared input '{{input.{name}}}' — declare it under inputs:"),
        )));
    };
    used.insert(name.to_string());
    Ok(spec)
}

/// `checkInputRefs`: every `{input.*}` reference must be declared; returns the
/// set of inputs that were referenced (stdin/prompt/params refs, `HWF_INPUT_*`
/// mentions in shell commands, agent-position refs).
///
/// # Errors
/// `WorkflowLoadError` on the first undeclared reference or agent-input misuse.
pub fn check_input_refs(
    file: &str,
    inputs: &[InputSpec],
    steps: &[FlatStep],
    agents: &HashSet<String>,
) -> Result<HashSet<String>, WorkflowLoadError> {
    let mut used = HashSet::new();
    for (idx, step) in steps.iter().enumerate() {
        match step {
            FlatStep::Shell { command, stdin } => {
                for name in text_input_refs(stdin.as_deref().unwrap_or("")) {
                    require(file, inputs, &name, idx, Some("stdin"), &mut used)?;
                }
                for spec in inputs {
                    if command.contains(&format!("HWF_INPUT_{}", spec.name)) {
                        require(file, inputs, &spec.name, idx, None, &mut used)?;
                    }
                }
            }
            FlatStep::Herdr { params, .. } => {
                for name in params_input_refs(params.as_ref()) {
                    require(file, inputs, &name, idx, Some("params"), &mut used)?;
                }
            }
            FlatStep::Agent { name, prompt, .. } => {
                for input_name in text_input_refs(prompt.as_deref().unwrap_or("")) {
                    require(file, inputs, &input_name, idx, Some("prompt"), &mut used)?;
                }
                if let Some(input_name) = agent_input_ref(name) {
                    let spec = require(file, inputs, &input_name, idx, Some("agent"), &mut used)?;
                    check_agent_input(file, idx, spec, agents)?;
                }
            }
            FlatStep::Open { .. } => {}
        }
    }
    Ok(used)
}
